src/options.ts

```diff
@@ -0,0 +1,125 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as minimist from 'minimist';
+import {resolve} from 'path';
+import {SignatureType, isValidSignatureType} from './types';
+
+/**
+ * Error thrown when an invalid option is provided.
+ */
+export class OptionsError extends Error {}
+
+/**
+ * Helper class for parsing an configurable option from provided CLI flags
+ * or environment variables.
+ */
+class ConfigurableOption<T> {
+  constructor(
+    public readonly cliOption: string,
+    private envVar: string,
+    private defaultValue: T,
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    private validator: (x: any) => T = x => x as T
+  ) {}
+
+  parse(cliArgs: minimist.ParsedArgs, envVars: NodeJS.ProcessEnv): T {
+    return this.validator(
+      cliArgs[this.cliOption] || envVars[this.envVar] || this.defaultValue
+    );
+  }
+}
+
+const PortOption = new ConfigurableOption('port', 'PORT', '8080');
+const FunctionTargetOption = new ConfigurableOption(
+  'target',
+  'FUNCTION_TARGET',
+  'function'
+);
+const SourceLocationOption = new ConfigurableOption(
+  'source',
+  'FUNCTION_SOURCE',
+  '',
+  resolve
+);
+const SignatureOption = new ConfigurableOption(
+  'signature-type',
+  'FUNCTION_SIGNATURE_TYPE',
+  'http' as SignatureType,
+  x => {
+    if (isValidSignatureType(x)) {
+      return x;
+    }
+    throw new OptionsError(
+      `Function signature type must be one of: ${SignatureType.join(', ')}.`
+    );
+  }
+);
```

Comment:
I feel like all of these options could just be put in an array/object, and not 4 constants. Then below you could just use that array directly.

Reply:
see comment below about difficulty of cleaning this up.